File: tests/tools/datasources_test.py
import json

from mcp_grafana.tools.datasources import (
    list_datasources,
    get_datasource_by_uid,
    get_datasource_by_name,
)

from . import mark_integration

# All tests in this file require a running Grafana instance and
# an asynchronous event loop.
pytestmark = mark_integration


async def test_list_datasources():
    datasources_bytes = await list_datasources()
    datasources = json.loads(datasources_bytes)
    assert len(datasources) == 1
    assert datasources[0]["name"] == "Robust Perception"
    assert datasources[0]["id"] == 1
    assert datasources[0]["uid"] == "robustperception"


async def test_get_datasource_by_uid():
    datasource_bytes = await get_datasource_by_uid("robustperception")
    datasource = json.loads(datasource_bytes)
    assert datasource["name"] == "Robust Perception"
    assert datasource["id"] == 1
    assert datasource["uid"] == "robustperception"


async def test_get_datasource_by_name():
    datasource_bytes = await get_datasource_by_name("Robust Perception")
    datasource = json.loads(datasource_bytes)
    assert datasource["name"] == "Robust Perception"
    assert datasource["id"] == 1
    assert datasource["uid"] == "robustperception"

File: tools/datasources_test.go
// Requires a Grafana instance running on localhost:3000,
// with a Prometheus datasource provisioned.
// Run with `go test -tags integration`.
//go:build integration

package tools

import (
	"context"
	"fmt"
	"net/url"
	"os"
	"testing"

	"github.com/go-openapi/strfmt"
	"github.com/grafana/grafana-openapi-client-go/client"
	mcpgrafana "github.com/grafana/mcp-grafana"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// newTestContext creates a new context with the Grafana URL and API key
// from the environment variables GRAFANA_URL and GRAFANA_API_KEY.
// TODO: move this to a shared file.
func newTestContext() context.Context {
	cfg := client.DefaultTransportConfig()
	cfg.Host = "localhost:3000"
	cfg.Schemes = []string{"http"}
	// Extract transport config from env vars, and set it on the context.
	if u, ok := os.LookupEnv("GRAFANA_URL"); ok {
		url, err := url.Parse(u)
		if err != nil {
			panic(fmt.Errorf("invalid %s: %w", "GRAFANA_URL", err))
		}
		cfg.Host = url.Host
		// The Grafana client will always prefer HTTPS even if the URL is HTTP,
		// so we need to limit the schemes to HTTP if the URL is HTTP.
		if url.Scheme == "http" {
			cfg.Schemes = []string{"http"}
		}
	}

	if apiKey := os.Getenv("GRAFANA_API_KEY"); apiKey != "" {
		cfg.APIKey = apiKey
	}

	client := client.NewHTTPClientWithConfig(strfmt.Default, cfg)
	return mcpgrafana.WithGrafanaClient(context.Background(), client)
}

func TestDatasourcesTools(t *testing.T) {
	t.Run("list datasources", func(t *testing.T) {
		ctx := newTestContext()
		result, err := listDatasources(ctx, ListDatasourcesParams{})
		require.NoError(t, err)
		// Three datasources are provisioned in the test environment (Prometheus and Loki).
		assert.Len(t, result, 3)
	})

	t.Run("list datasources for type", func(t *testing.T) {
		ctx := newTestContext()
		result, err := listDatasources(ctx, ListDatasourcesParams{Type: "Prometheus"})
		require.NoError(t, err)
		// Only two Prometheus datasources are provisioned in the test environment.
		assert.Len(t, result, 2)
	})

	t.Run("get datasource by uid", func(t *testing.T) {
		ctx := newTestContext()
		result, err := getDatasourceByUID(ctx, GetDatasourceByUIDParams{
			UID: "prometheus",
		})
		require.NoError(t, err)
		assert.Equal(t, "Prometheus", result.Name)
	})

	t.Run("get datasource by name", func(t *testing.T) {
		ctx := newTestContext()
		result, err := getDatasourceByName(ctx, GetDatasourceByNameParams{
			Name: "Prometheus",
		})
		require.NoError(t, err)
		assert.Equal(t, "Prometheus", result.Name)
	})
}
